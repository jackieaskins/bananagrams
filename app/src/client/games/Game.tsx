import { Box, Flex, useBreakpointValue } from "@chakra-ui/react";
import Konva from "konva";
import {
  useCallback,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from "react";
import GameFooter from "./GameFooter";
import GameSidebar from "./GameSidebar";
import Canvas from "@/client/canvas/Canvas";
import { CanvasContext } from "@/client/canvas/CanvasContext";
import KeysProvider from "@/client/keys/KeysProvider";
import { useNavMenu } from "@/client/menus/NavMenuContext";
import SelectedTilesProvider from "@/client/tiles/SelectedTilesProvider";

const TILE_SIZE = 32;

export default function Game(): JSX.Element {
  const gameSidebarRef = useRef<HTMLDivElement>(null);
  const gameBarRef = useRef<HTMLDivElement>(null);

  const [sidebarExpanded, setSidebarExpanded] = useState(true);
  const [hasManuallyToggledSidebar, setHasManuallyToggledSidebar] =
    useState(false);
  const shouldDefaultSidebarOpen = useBreakpointValue(
    { base: true, sm: false, md: true },
    { fallback: "md" },
  );
  const expandSidebar = useCallback((expanded: boolean) => {
    setSidebarExpanded(expanded);
    setHasManuallyToggledSidebar(true);
  }, []);
  useEffect(() => {
    if (!hasManuallyToggledSidebar) {
      setSidebarExpanded(shouldDefaultSidebarOpen!);
    }
  }, [hasManuallyToggledSidebar, shouldDefaultSidebarOpen]);

  const [cursorPosition, setCursorPosition] = useState({ x: 0, y: 0 });
  const [offset, setOffset] = useState({ x: 0, y: 0 });
  const stageRef = useRef<Konva.Stage>(null);
  const [size, setSize] = useState({
    width: window.innerWidth,
    height: window.innerHeight,
  });

  const [, setRenderNavMenu] = useNavMenu();
  useEffect(() => {
    setRenderNavMenu(false);

    return () => {
      setRenderNavMenu(true);
    };
  }, [setRenderNavMenu]);

  useLayoutEffect(() => {
    const currGameBarRef = gameBarRef.current;
    const currGameSidebarRef = gameSidebarRef.current;

    function updateSize() {
      setSize({
        width: window.innerWidth - (currGameSidebarRef?.clientWidth ?? 0),
        height: window.innerHeight - (currGameBarRef?.clientHeight ?? 0),
      });
    }
    const observer = new ResizeObserver(updateSize);

    window.addEventListener("resize", updateSize);
    if (currGameBarRef) observer.observe(currGameBarRef);
    if (currGameSidebarRef) observer.observe(currGameSidebarRef);

    updateSize();

    return () => {
      window.removeEventListener("resize", updateSize);
      observer.disconnect();
    };
  }, []);

  return (
    <Flex height="100vh" width="100vw" direction="row">
      <Box>
        <CanvasContext.Provider
          value={{
            cursorPosition,
            offset,
            playable: true,
            size,
            stageRef,
            tileSize: TILE_SIZE,
          }}
        >
          <SelectedTilesProvider>
            <KeysProvider>
              <Canvas
                setOffset={setOffset}
                setCursorPosition={setCursorPosition}
              />
            </KeysProvider>
          </SelectedTilesProvider>
        </CanvasContext.Provider>

        <Box ref={gameBarRef}>
          <GameFooter sidebarExpanded={sidebarExpanded} />
        </Box>
      </Box>

      <Box ref={gameSidebarRef}>
        <GameSidebar expanded={sidebarExpanded} setExpanded={expandSidebar} />
      </Box>
    </Flex>
  );
}
