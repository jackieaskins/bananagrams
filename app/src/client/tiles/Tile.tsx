import { Box, Text } from "@chakra-ui/react";
import { useDrag } from "react-dnd";
import { Tile as TileType } from "@/types/tile";

type TileProps = {
  boardLocation: { x: number; y: number } | null;
  tile: TileType;
  color?: string;
};

export default function Tile({
  boardLocation,
  tile: { id, letter },
  color = "black",
}: TileProps): JSX.Element {
  const [{ isDragging }, dragRef] = useDrag({
    item: { type: "TILE", id, boardLocation },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  const margin = boardLocation ? "0" : "5px";

  return (
    <Box
      ref={dragRef}
      alignItems="center"
      backgroundColor="yellow.100"
      border="1px"
      borderColor={color}
      borderRadius="5px"
      color={color}
      cursor="move"
      display="inline-flex"
      height="25px"
      justifyContent="center"
      margin={margin}
      opacity={isDragging ? 0.5 : 1}
      width="25px"
    >
      <Text as="b">{letter}</Text>
    </Box>
  );
}
