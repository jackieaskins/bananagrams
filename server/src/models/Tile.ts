import BaseModel from './BaseModel';

export type TileJSON = {
  id: string;
  letter: string;
};

export default class Tile implements BaseModel<TileJSON> {
  private id: string;
  private letter: string;

  constructor(id: string, letter: string) {
    this.id = id;
    this.letter = letter;
  }

  getId(): string {
    return this.id;
  }

  getLetter(): string {
    return this.letter;
  }

  toJSON(): TileJSON {
    const { id, letter } = this;
    return { id, letter };
  }

  reset(): void {
    return;
  }
}
