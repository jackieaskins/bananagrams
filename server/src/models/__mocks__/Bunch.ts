export default jest.fn().mockImplementation(() => ({
  toJSON: jest.fn().mockReturnValue('BunchJSON'),
  reset: jest.fn(),
}));
