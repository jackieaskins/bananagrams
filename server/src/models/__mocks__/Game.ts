import Player from '../Player';

export default jest.fn().mockImplementation(() => {
  const players: Player[] = [];

  return {
    addPlayer: jest.fn().mockImplementation((player) => {
      players.push(player);
    }),
    getPlayers: jest.fn().mockReturnValue(players),
  };
});
