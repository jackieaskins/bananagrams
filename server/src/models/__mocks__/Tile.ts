export default jest.fn().mockImplementation((id, letter) => ({
  getId: jest.fn().mockReturnValue(id),
  getLetter: jest.fn().mockReturnValue(letter),
  toJSON: jest.fn().mockReturnValue(`TileJSON - ${id} ${letter}`),
}));
