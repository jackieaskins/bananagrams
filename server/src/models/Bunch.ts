import Tile, { TileJSON } from './Tile';
import tileBreakdown from '../tileBreakdown';
import BaseModel from './BaseModel';
import Game from './Game';

export type BunchJSON = TileJSON[];

const MULTIPLIER_DIVISOR = 4;

export default class Bunch implements BaseModel<BunchJSON> {
  private game: Game;
  private tiles: Tile[] = [];

  constructor(game: Game) {
    this.game = game;
  }

  getTiles(): Tile[] {
    return [...this.tiles];
  }

  toJSON(): BunchJSON {
    return this.tiles.map((tile) => tile.toJSON());
  }

  reset(): void {
    if (this.game.isShortenedGame()) {
      this.tiles = [
        new Tile('A1', 'A'),
        new Tile('E1', 'E'),
        new Tile('T1', 'T'),
      ];

      return;
    }

    const multiplier = Math.ceil(
      this.game.getPlayers().length / MULTIPLIER_DIVISOR
    );

    this.tiles = tileBreakdown
      .map(({ letter, count }) =>
        Array(count * multiplier)
          .fill(null)
          .map((_, i) => new Tile(`${letter}${i}`, letter))
      )
      .flat();
  }

  addTiles(tiles: Tile[]): void {
    this.tiles = [...this.tiles, ...tiles];
  }

  removeTiles(count: number): Tile[] {
    if (this.tiles.length < count) {
      throw new Error(
        `The bunch has less than ${count} ${count === 1 ? 'tile' : 'tiles'}`
      );
    }

    return Array.from(Array(count)).map(() => {
      const index = Math.floor(Math.random() * this.tiles.length);

      const tile = this.tiles[index];
      this.tiles = [
        ...this.tiles.slice(0, index),
        ...this.tiles.slice(index + 1),
      ];

      return tile;
    });
  }
}
