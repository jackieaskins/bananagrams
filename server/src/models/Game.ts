import Bunch, { BunchJSON } from './Bunch';
import Player, { PlayerJSON } from './Player';
import BaseModel from './BaseModel';

export type Snapshot = PlayerJSON[] | null;
export type GameJSON = {
  gameId: string;
  gameName: string;
  isInProgress: boolean;
  bunch: BunchJSON;
  players: PlayerJSON[];
  previousSnapshot: Snapshot;
};

export default class Game implements BaseModel<GameJSON> {
  private id: string;
  private name: string;
  private inProgress = false;
  private bunch: Bunch = new Bunch();
  private players: Player[] = [];
  private previousSnapshot: Snapshot = null;

  constructor(id: string, name: string) {
    this.id = id;
    this.name = name;
  }

  getId(): string {
    return this.id;
  }

  getName(): string {
    return this.name;
  }

  isInProgress(): boolean {
    return this.inProgress;
  }

  setInProgress(inProgress: boolean): void {
    this.inProgress = inProgress;
  }

  getBunch(): Bunch {
    return this.bunch;
  }

  getPlayers(): Player[] {
    return [...this.players];
  }

  getSnapshot(): Snapshot {
    return this.previousSnapshot ? [...this.previousSnapshot] : null;
  }

  setSnapshot(previousSnapshot: PlayerJSON[]): void {
    this.previousSnapshot = [...previousSnapshot];
  }

  reset(): void {
    this.bunch.reset();
    this.players.forEach((player) => player.reset());
  }

  toJSON(): GameJSON {
    const {
      id: gameId,
      name: gameName,
      inProgress: isInProgress,
      bunch,
      players,
      previousSnapshot,
    } = this;

    return {
      gameId,
      gameName,
      isInProgress,
      bunch: bunch.toJSON(),
      players: players.map((player) => player.toJSON()),
      previousSnapshot,
    };
  }

  addPlayer(player: Player): void {
    this.players = [...this.players, player];
  }

  removePlayer(userId: string): void {
    this.players = this.players.filter(
      (player) => player.getUserId() !== userId
    );
  }
}
