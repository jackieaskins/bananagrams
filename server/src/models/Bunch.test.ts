import Bunch from './Bunch';
import Game from './Game';
import Tile from './Tile';

jest.mock('../tileBreakdown', () => [
  { letter: 'A', count: 2 },
  { letter: 'B', count: 5 },
]);

describe('Bunch Model', () => {
  let game: Game;
  let bunch: Bunch;

  beforeEach(() => {
    game = new Game('gameId', 'gameName');
    bunch = new Bunch(game);
  });

  describe('getTiles', () => {
    it('returns empty array by default', () => {
      expect(bunch.getTiles()).toHaveLength(0);
    });
  });

  describe('toJSON', () => {
    it('returns fields converted to JSON', () => {
      bunch.addTiles([new Tile('A1', 'A')]);
      expect(bunch.toJSON()).toMatchSnapshot();
    });
  });

  describe('reset', () => {
    it('creates bunch with 3 tiles if shortened game', () => {
      const shortenedGame = new Game('gameId', 'gameName', true);
      const shortenedBunch = new Bunch(shortenedGame);

      shortenedBunch.reset();

      expect(shortenedBunch.getTiles()).toMatchSnapshot();
    });

    it('generates tiles with multiplier of 1 when <= 4 players', () => {
      jest.spyOn(game, 'getPlayers').mockReturnValue(Array(4).fill(null));

      bunch.reset();

      const tiles = bunch.getTiles();
      expect(tiles).toMatchSnapshot();
      expect(tiles).toHaveLength(7);
    });

    it('generates tiles with multiplier of 2 when > 4 players', () => {
      jest.spyOn(game, 'getPlayers').mockReturnValue(Array(7).fill(null));

      bunch.reset();

      const tiles = bunch.getTiles();
      expect(tiles).toMatchSnapshot();
      expect(tiles).toHaveLength(14);
    });
  });

  describe('addTiles', () => {
    it('adds tiles to bunch', () => {
      const tiles = [new Tile('A1', 'A'), new Tile('B1', 'B')];
      bunch.addTiles(tiles);

      expect(bunch.getTiles()).toEqual(tiles);
    });
  });

  describe('removeTiles', () => {
    const tileA1 = new Tile('A1', 'A');
    const tileB1 = new Tile('B1', 'B');
    const tileC1 = new Tile('C1', 'C');
    const tiles = [tileA1, tileB1, tileC1];

    beforeAll(() => {
      jest.spyOn(global.Math, 'random').mockReturnValue(0);
    });

    afterAll(() => {
      jest.spyOn(global.Math, 'random').mockRestore();
    });

    it('throws an error if there is less than 1 requested tile', () => {
      expect(() => bunch.removeTiles(1)).toThrowErrorMatchingSnapshot();
    });

    it('throws an error if there are less than number of requested tiles', () => {
      bunch.addTiles(tiles);
      expect(() => bunch.removeTiles(4)).toThrowErrorMatchingSnapshot();
    });

    it('returns number of tiles requested', () => {
      bunch.addTiles(tiles);
      expect(bunch.removeTiles(2)).toEqual([tileA1, tileB1]);
    });
  });
});
