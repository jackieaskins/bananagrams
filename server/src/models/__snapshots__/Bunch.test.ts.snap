// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bunch Model removeTiles throws an error if there are less than number of requested tiles 1`] = `"The bunch has less than 4 tiles"`;

exports[`Bunch Model removeTiles throws an error if there is less than 1 requested tile 1`] = `"The bunch has less than 1 tile"`;

exports[`Bunch Model reset generates tiles with multiplier of 1 when <= 4 players 1`] = `
Array [
  Object {
    "id": "A0",
    "letter": "A",
  },
  Object {
    "id": "A1",
    "letter": "A",
  },
  Object {
    "id": "B0",
    "letter": "B",
  },
  Object {
    "id": "B1",
    "letter": "B",
  },
  Object {
    "id": "B2",
    "letter": "B",
  },
  Object {
    "id": "B3",
    "letter": "B",
  },
  Object {
    "id": "B4",
    "letter": "B",
  },
]
`;

exports[`Bunch Model reset generates tiles with multiplier of 2 when > 4 players 1`] = `
Array [
  Object {
    "id": "A0",
    "letter": "A",
  },
  Object {
    "id": "A1",
    "letter": "A",
  },
  Object {
    "id": "A2",
    "letter": "A",
  },
  Object {
    "id": "A3",
    "letter": "A",
  },
  Object {
    "id": "B0",
    "letter": "B",
  },
  Object {
    "id": "B1",
    "letter": "B",
  },
  Object {
    "id": "B2",
    "letter": "B",
  },
  Object {
    "id": "B3",
    "letter": "B",
  },
  Object {
    "id": "B4",
    "letter": "B",
  },
  Object {
    "id": "B5",
    "letter": "B",
  },
  Object {
    "id": "B6",
    "letter": "B",
  },
  Object {
    "id": "B7",
    "letter": "B",
  },
  Object {
    "id": "B8",
    "letter": "B",
  },
  Object {
    "id": "B9",
    "letter": "B",
  },
]
`;

exports[`Bunch Model toJSON returns fields converted to JSON 1`] = `
Array [
  Object {
    "id": "A1",
    "letter": "A",
  },
]
`;
